kyo-core/shared/src/main/scala/kyo/StreamCoreExtensions.scala

```diff
@@ -304,80 +304,59 @@ object StreamCoreExtensions:
-                                                            Async.run {
-                                                                nextChunkEffect.map: nextChunk =>
-                                                                    prevChunkFiber.get.map: prevChunk =>
-                                                                        prevEmitFiber.get.andThen:
-                                                                            val chunk = prevChunk ++ nextChunk
-                                                                            channel.put(Present(chunk)).andThen:
-                                                                                parAdjustmentRef.updateAndGet(_ + chunk.size).unit
-                                                            }.map: nextFiber =>
-                                                                Loop.done(Loop.continue(
-                                                                    (nextFiber, newRemainingPar),
-                                                                    cont(())
-                                                                ))
-                                                        else
-                                                            bug("Illegal state: there is remaining parallel and remaining chunk in mapPar")
-                                                        end if
-                                    ,
-                                    done = {
-                                        case ((lastFiber, _), _) =>
-                                            lastFiber.get.andThen:
-                                                channel.put(Absent)
-                                    }
-                                )
-
-                                Abort.fold[E | Closed](
-                                    onSuccess = _ => Abort.run(channel.put(Absent)).unit,
-                                    onFail = {
-                                        case _: Closed       => bug("buffer closed unexpectedly")
-                                        case e: E @unchecked => Abort.run(channel.put(Absent)).andThen(Abort.fail(e))
-                                    },
-                                    onPanic = e => Abort.run(channel.put(Absent)).andThen(Abort.panic(e))
-                                )(handledStream)
-
-                            background.map: backgroundFiber =>
-                                emitMaybeChunksFromChannel(channel).andThen:
-                                    backgroundFiber.get.unit
-                }
+                Channel.initWith[Maybe[Chunk[V2]]](bufferSize): channelOut =>
+                    // Staging channel acts as concurrency limiter: contains fibers that either contain a
+                    // chunk to be published, or a signal to continue or end
+                    Channel.initWith[Fiber[E | Closed, Boolean | Chunk[V2]]](parallel - 1): stagingChannel =>
```

Comment:
no need to change it but wouldn't a `Meter` work well for this?

Reply:
The problem with `Meter` is ensuring that the end-of-stream signal is not published until all pending transformations have published.